relay-server/src/actors/upstream.rs

```diff
@@ -284,43 +319,114 @@ impl UpstreamRelay {
     pub fn new(config: Arc<Config>) -> Self {
         UpstreamRelay {
             backoff: RetryBackoff::new(config.http_max_retry_interval()),
+            network_status_backoff: RetryBackoff::new(config.http_max_retry_interval()),
             max_inflight_requests: config.max_concurrent_requests(),
             auth_state: AuthState::Unknown,
             num_inflight_requests: 0,
             high_prio_requests: VecDeque::new(),
             low_prio_requests: VecDeque::new(),
             first_error: None,
+            authentication_denied: false,
+            is_waiting_to_authenticate: false,
             config,
         }
     }
 
+    /// Predicate, checks if a Relay performs authentication
+    fn authenticates(&self) -> bool {
+        // only managed mode relays perform authentication
+        self.config.relay_mode() == RelayMode::Managed
+    }
+
+    /// Predicate, checks if a Relay does re-authentication
+    fn re_authenticates(&self) -> bool {
```

Comment:
You use this function to check if reauthentication should be performed, only to check for the auth interval again. I would propose you combine this into a single function that has the following implementation:

```rust
fn auth_interval(&self) -> Option<Duration> {
  if self.authentication_denied {
    None
  } else {
    self.config.http_auth_interval()
  }
}
```

_nit_: Note that `http_auth_interval` already checks the processing mode, and I would propose to move the check for the relay mode there as well, so that there is a single place where all these "static" checks are being performed.

Reply:
I agree with you that there is one place (from 2) where there is effectively a duplicate check for `http_auth_interval.is_some()` but I think it is much cleaner like that than having auth_interval encapsulate authentication logic. 

If somebody looks at the code ( at least if I look at the code) and see  should_authenticate() it is quite clear what happens in the function... if I se auth_interval() I don't expect that it has some business logic  beyond getting the setting.